messenger.go

```diff
@@ -697,3 +724,129 @@ func (m *Messenger) MarkMessagesSeen(ids ...string) error {
 	return m.persistence.UpdateMessageOutgoingStatus(id, newOutgoingStatus)
 }
+
+// postProcessor performs a set of actions on newly retrieved messages.
+// If persist is true, it saves the messages into the database.
+// If matchChat is true, it matches each messages against a Chat instance.
+type postProcessor struct {
+	myPublicKey *ecdsa.PublicKey
+	persistence *sqlitePersistence
+	logger      *zap.Logger
+
+	config postProcessorConfig
+}
+
+type postProcessorConfig struct {
+	MatchChat bool // match each messages to a chat; may result in a new chat creation
+	Persist   bool // if true, all sent and received user messages will be persisted
+}
+
+func newPostProcessor(m *Messenger, config postProcessorConfig) *postProcessor {
+	return &postProcessor{
+		myPublicKey: &m.identity.PublicKey,
+		persistence: m.persistence,
+		logger:      m.logger,
+		config:      config,
+	}
+}
+
+func (p *postProcessor) Run(messages []*protocol.Message) ([]*protocol.Message, error) {
+	var err error
+
+	p.logger.Debug("running post processor")
+
+	var fns []func([]*protocol.Message) ([]*protocol.Message, error)
+
+	// Order is important. Persisting messages should be always at the end.
+	if p.config.MatchChat {
+		fns = append(fns, p.matchMessages)
+	}
+	if p.config.Persist {
+		fns = append(fns, p.saveMessages)
+	}
+
+	for _, fn := range fns {
+		messages, err = fn(messages)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return messages, nil
+}
+
+func (p *postProcessor) saveMessages(messages []*protocol.Message) ([]*protocol.Message, error) {
+	_, err := p.persistence.SaveMessages(messages)
+	if err != nil {
+		return nil, err
+	}
+	return messages, nil
+}
+
+func (p *postProcessor) matchMessages(messages []*protocol.Message) ([]*protocol.Message, error) {
+	chats, err := p.persistence.Chats()
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]*protocol.Message, 0, len(messages))
+	for _, message := range messages {
+		chat, err := p.matchMessage(message, chats)
+		if err != nil {
+			p.logger.Error("failed to match a chat to a message", zap.Error(err))
+			continue
+		}
+		message.ChatID = chat.ID
+		result = append(result, message)
+	}
+	return result, nil
+}
+
+func (p *postProcessor) matchMessage(message *protocol.Message, chats []*Chat) (*Chat, error) {
+	switch {
+	case message.MessageT == protocol.MessageTypePublicGroup:
+		// For public messages, all outgoing and incoming messages have the same chatID
+		// equal to a public chat name.
+		chatID := message.Content.ChatID
+		chat := findChatByID(chatID, chats)
+		if chat == nil {
+			return nil, errors.New("received a public message from non-existing chat")
+		}
+		return chat, nil
+	case message.MessageT == protocol.MessageTypePrivate && isPubKeyEqual(message.SigPubKey, p.myPublicKey):
+		// It's a private message coming from us so we rely on Message.Content.ChatID.
+		chatID := message.Content.ChatID
+		chat := findChatByID(chatID, chats)
```

Comment:
Not sure this is correct, the behavior should be that if the chat does not exist it should be created in this case. Say I send a message to you from device A, I will send the same message to device B. Device B should create a chat similarly to the next case.